custom_components/xiaomi_miio_fan/fan.py

```diff
@@ -2101,3 +2133,283 @@ def set_rotate(self, direction: FanMoveDirection):
         elif direction == FanMoveDirection.Right:
             value = 2
         return self.set_property("motor_control", value)
+
+class XiaomiFanP39(XiaomiFanMiot):
+    """Representation of a Xiaomi Fan P39."""
+
+    def __init__(self, name, device, model, unique_id, retries, preset_modes_override):
+        """Initialize the fan entity."""
+        super().__init__(name, device, model, unique_id, retries, preset_modes_override)
+
+        self._device_features = FEATURE_FLAGS_FAN_P39
+        self._available_attributes = AVAILABLE_ATTRIBUTES_FAN_P39
+        self._percentage = None
+        self._preset_modes = list(FAN_PRESET_MODES_P39)
+        if preset_modes_override is not None:
+            self._preset_modes = preset_modes_override
+
+        self._preset_mode = None
+        self._oscillate = None
+        self._natural_mode = False
+
+        self._state_attrs.update(
+            {attribute: None for attribute in self._available_attributes}
+        )
+
+    @property
+    def supported_features(self) -> int:
+        return (
+            SUPPORT_DIRECTION
+            | SUPPORT_OSCILLATE
+            | SUPPORT_PRESET_MODE
+            | SUPPORT_SET_SPEED
+        )
+
+    """
+    TODO:
+    - fan.turn_on works, but doesn't go to the declared percentage. This is
```

Comment:
Do you address these issues before merging?

Reply:
Cleaned up TODOs in latest commits.
Regarding child lock: as stated, this is accurately reflected in HA but I can't seem to be able to actually change it. Am I missing something? Is there supposed to be a button?

Reply:
There is a service called `xiaomi_miio_fan.fan_set_child_lock_on` and `xiaomi_miio_fan.fan_set_child_lock_off`. Did you try to call the service to change the setting?

Reply:
Yes, that worked.
So the functionality itself is intact. I was just looking for a dedicated button.